### PR TITLE
lexical: init at 0.5.2

### PR DESCRIPTION
**Pull request by GaetanLepage**
## Description of changes

Add the [lexical](https://github.com/lexical-lsp/lexical) Elixir language server.

Note: I have no experience in packaging things for this language.
I use the [project's own flake](https://github.com/lexical-lsp/lexical/blob/main/nix/lexical.nix) as a starting point.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [x] aarch64-linux
  - [x] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: pkgs/by-name/le/lexical/package.nix [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    owner = "lexical-lsp";
+    repo = "lexical";
+    rev = "refs/tags/v${version}";
+    hash = "sha256-HWqwJ7PAz80bm6YeDG84hLWPE11n06K98GOyeDQWZWU=";
+  };
+
+  mixFodDeps = beamPackages.fetchMixDeps {
+    inherit pname version src;
+
+    sha256 = "sha256-G0mT+rvXZWLJIMfrhxq3TXt26wDImayu44wGEYJ+3CE=";
```
Comment (adamcstephens) [inline pkgs/by-name/le/lexical/package.nix:22]:
```suggestion
    hash = "sha256-G0mT+rvXZWLJIMfrhxq3TXt26wDImayu44wGEYJ+3CE=";
```

File: pkgs/by-name/le/lexical/package.nix [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+  '';
+
+
+  meta = with lib; {
+    description = "Lexical is a next-generation elixir language server";
+    homepage = "https://github.com/lexical-lsp/lexical";
+    license = licenses.asl20;
+    maintainers = with maintainers; [ GaetanLepage ];
+    mainProgram = "lexical";
+    platforms = platforms.all;
```
Comment (adamcstephens) [inline pkgs/by-name/le/lexical/package.nix:54]:
```suggestion
    platforms = beamPackages.erlang.meta.platforms;
```

File: pkgs/by-name/le/lexical/package.nix [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+      true
+    '';
+  in ''
+    substituteInPlace "$out/bin/start_lexical.sh" --replace 'elixir_command=' 'elixir_command="${elixir}/bin/"'
+    rm "$out/bin/activate_version_manager.sh"
+    ln -s ${activate_version_manager} "$out/bin/activate_version_manager.sh"
+
+    mv "$out/bin" "$out/binsh"
+
+    makeWrapper "$out/binsh/start_lexical.sh" "$out/bin/lexical" --set RELEASE_COOKIE lexical
```
Comment (adamcstephens) [inline pkgs/by-name/le/lexical/package.nix:44]:
```suggestion
  postInstall = ''
    substituteInPlace "$out/bin/start_lexical.sh" --replace 'elixir_command=' 'elixir_command="${elixir}/bin/"'

    mv "$out/bin" "$out/libexec"

    makeWrapper "$out/libexec/start_lexical.sh" "$out/bin/lexical" --set RELEASE_COOKIE lexical
```

Upstream could maybe clean this up some?
https://github.com/lexical-lsp/lexical/blob/bb1c1e3f66160653f95bd05af7589297233fec96/bin/start_lexical.sh#L7

`libexec` is a bit more standard than `binsh`, so better to use this. I considered getting rid of the extra directory completely, but I like only exposing `bin/lexical` to users.

**Review by adamcstephens — CHANGES_REQUESTED** [review]
I'd encourage you to use `nixfmt-rfc-style` to format this new package up front.

File: pkgs/by-name/le/lexical/package.nix (hunk 3 above)
Comment (GaetanLepage) [reply pkgs/by-name/le/lexical/package.nix]: I made the change.

**Comment by GaetanLepage** [comment]
> I'd encourage you to use `nixfmt-rfc-style` to format this new package up front.

Done !

**Comment by GaetanLepage** [comment]
Thanks for the review @adamcstephens !

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 300802` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lexical</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 300802` run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lexical</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 300802` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lexical</li>
  </ul>
</details>

**Comment by GaetanLepage** [comment]
Result of `nixpkgs-review pr 300802` run on aarch64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>lexical</li>
  </ul>
</details>

**Review by adamcstephens — APPROVED** [review]

**Pull request merged**